buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -565,10 +566,59 @@ static ByteBuf encodeString0(ByteBufAllocator alloc, boolean enforceHeap, CharBu
         }
     }
 
+    public static String readUtf8(ByteBuf buf, int readerIndex, int length) {
+        if (buf instanceof AbstractByteBuf) {
+            return readUtf8Fast((AbstractByteBuf) buf, readerIndex, length);
+        } else {
+            return readUtf8Slow(buf, readerIndex, length);
+        }
+    }
+
+    private static String readUtf8Slow(ByteBuf buf, int readerIndex, int length) {
+        if (buf instanceof WrappedByteBuf) {
+            ByteBuf wrapped;
+            while ((wrapped = buf.unwrap()) != null) {
+                buf = wrapped;
+            }
+            if (buf instanceof AbstractByteBuf) {
+                return readUtf8Fast((AbstractByteBuf) buf, readerIndex, length);
+            }
+        }
+        UTF8Decoder.UTF8Processor processor = new UTF8Decoder.UTF8Processor(new char[length]);
+        buf.forEachByte(processor);
+        return processor.toString();
+    }
+
+    // Fast-Path for a AbstractByteBuf
+    static String readUtf8Fast(AbstractByteBuf buf, int readerIndex, int length) {
+        buf.checkIndex(readerIndex, length);
+        UTF8Decoder.UTF8Processor processor = new UTF8Decoder.UTF8Processor(new char[length]);
+        /*
+         * Inline AbstractByteBuf.forEachByte by hand since the JIT doesn't do it it properly
```

Comment:
If the method becomes "hot enough" it should be inlined (possibly even if it is over the [-XX:MaxInlineSize](https://docs.oracle.com/javase/8/docs/technotes/tools/unix/java.html) limit). This "hot enough" is typically defined as a heuristic based upon method execution count and method byte code size. Can you provide the output of `-XX:+PrintInlining` (relevant to `forEach*`), the code used to generate this output, and the JVM arguments?